Tier two network sync new architecture, regtest support + MN activation functional test.

furszy opened the pull request:
-- Only running on regtest --

Covering the following points:
#### 1) A new synchronization structure for the tier two network.
It has no redundant checks nor redundant requests (there by less network bloat, faster synchronization) and the code is by far more readable and maintainable than the previous version.

#### 2) Tier two network regtest support.
Adding Masternodes and Governance capabilities to the regtest network.

#### 3) Sync and masternode activation test.
Functional test validating the correct operation of the tier two network sync and the masternode broadcasting & activation process.

#### TODO:
* Clean peer state on peer disconnection (connect signal)
* Spam filter + spam functional test.
* Not responding peers timeout test + spam functional test.
* Sporks broadcasting end message (currently it's using the SPORK_INVALID as end message).
* Data verification from several peers (currently running, only need to implement a sync threshold to move to the next level) + spam functional test.
* Further decoupling from the old structure.
* Previous version cleanup.

------------
Have pushed it under WIP to be able to test other great PRs like #1826.

furszy commented:
Added a commit expanding the test coverage for the following scenarios: masternode expiration, re initialization after expiration, removal, re initialization after removal and collateral spent.

This will give us higher security for #1826 and #1835 testing on the tier two modifications (hopefully speeding up the review as well).

Time wise: the test will take some more time due the different protocol rules/restrictions to configure the masternode, pass from one state to another, setup the node, shut it down etc. As manual testing sucks, and consumes the entire day (when you have to setup a whole network and enable a group of masternodes all by yourself..), don't care on how much time this functional test is taking, it's hours and hours that we going to win :) .

furszy commented:
long night :) , added another commit increasing further the test coverage.
 
First budget/governance area functional test, covering the following points:

 1) Group of masternodes setup/creation.
 2) Proposal creation.
 3) Vote creation.
 4) Proposal and vote broadcast.
 5) Proposal and vote sync.


Pending for future work (if anyone want, feel more than welcome to tackle any of the following points or code new ones on top of this PR):

Add test coverage for budget finalization, proposals being funded or not, proposals removal, budget projection, change vote.

Future future plus: it would be great if we can prepare a framework setup for the governance tests. Start with N amount of masternodes instead of have to set & initialize them in each test, maybe with proposals in the network too etc.

furszy commented:
rebased on master

random-zebra commented on test/functional/masternode_activation.py:
```diff
@@ -0,0 +1,226 @@
+        time.sleep(20)
+        self.check_mnsync_finished()
+
+    def satoshi_round(amount):
+        return Decimal(amount).quantize(Decimal('0.00000001'), rounding=ROUND_DOWN)
```
no need to duplicate/reimplement this. Just import it from util.py

furszy replied:
updating

random-zebra commented on src/masternode-sync.h:
```diff
@@ -24,6 +29,11 @@
 class CMasternodeSync;
 extern CMasternodeSync masternodeSync;
 
+struct TierTwoPeerData {
+    // map of message --> last request timestamp, bool hasResponseArrived.
+    std::map<const char*, std::pair<int64_t, bool>> mapMsgData;
+};
```
instead of a struct with a single variable, it would be cleaner with a typedef.
This way all `peerData.mapMsgData` would be just `peerData`.

furszy replied:
I left it as a struct because will add other members moving forward with the new syncing process. This is a naive regtest-only sync.
It basically will mean less changes in the upcoming work.

random-zebra reviewed (commented):
Code review ACK, looking pretty good.
Aside from a couple notes left inline, have two minor talking points:
-  I'd rather remove commits 594fa7a924d028444687c789083c9e997385bdfd and 6f164f6775588a474225c0a50798694ae694f674 (and most `sleep` in the test).
Time constants can remain the same as mainnet (if you use mocktime).
- `LogPrintf` in `ProcessSporkMsg` could be very spammy. Maybe you could move to `LogPrint` with `BCLog::NET` or some other category.

furszy commented:
> I'd rather remove commits 594fa7a and 6f164f6 (and most sleep in the test).
Time constants can remain the same as mainnet (if you use mocktime).

i played with the mocktime implementation for a while and agree that would be cleaner with it but.. there still a missing piece there. MNs weren't getting enabled.

Better to merge this initial step and continue moving forward on top. So neither you or me are depending on each other's work and can continue moving on.

And btw, i actually think that 594fa7a and 6f164f6 are good outside of the mock time changes. I'm more on the side of keep them with a slightly modification in the upcoming work: 
agree on removing the mainnet/regtest times division with the mock time changes (this will mean only touch `masternode.cpp`), keeping the static functions call migration.
So in a future, we can implement, a little bit easier, a `TierTwoConsensusParams` similar concept and not have all consensus constant definitions all around the tier two sources.

> LogPrintf in ProcessSporkMsg could be very spammy. Maybe you could move to LogPrint with BCLog::NET or some other category.

Good catch 👍 , moving them.

furszy commented:
Updated per feedback.

furszy commented:
Rebased on master with some changes:

1) replaced the ping update hack for #1886 correction.
2) migrated `masternodes_governance_basic.py` to use mock time.
3) renamed functional tests to be using the `tiertwo_xxx` prefix.

For me, it's good to go in this way (merging 1886 first). It's already fulfilling a lot of needs.

furszy commented:
Rebased on top of master after #1886 merge.

Plus added two more commits to verify the budget finalization sync process, fulfilling the "governance_sync_basic" test purpose of validating the correct functioning of the tier two messages sync process.
With this changes, we are able to merge #1845 with some minimum checks for the synchronization process.

Ready for review and merge. Will not add any other change here.

random-zebra commented on test/functional/tiertwo_governance_sync_basic.py:
```diff
@@ -159,12 +175,17 @@ def check_proposal_existence(self, proposalName, proposalHash):
+                if (voteInfo["mnId"] == mnCollateralHash) :
+                    assert_equal(voteInfo["Vote"], voteType)
+                    found = True
+            assert_true(found, "Error checking vote existence in node " + str(i))
```
nit: could rewrite all this function in a single pythonic one-liner as:
```python
assert_equal([x["Vote"] for node in self.nodes for x in node.getbudgetvotes(proposalName)
                      if x["mnId"] == mnCollateralHash], [voteType] * len(self.nodes))
```

furszy replied:
would be nice if we could print in this magic line which node wasn't able to sync as well.

random-zebra commented on test/functional/masternodes_governance_basic.py:
```diff
@@ -0,0 +1,282 @@
+            if (mnData["txhash"] == mnTxHash):
+                assert_equal(mnData["status"], status)
+                found = True
+        assert(found)
```
nit: could rewrite all this function in a single pythonic one-liner as:
```python
assert_equal(len([x for x in node.listmasternodes() if x["txhash"] == mnTxHash]), 1)
```

random-zebra reviewed (approved):
Awesome first step :tumbler_glass: very much needed.
Left few more comments (mostly around the use of mocktime which seems a bit funky), but they can be addressed in a later PR, so ACK 9a1e5e29317796c93bdef6c2df21ef50e6c68580

Fuzzbawls reviewed (approved):
ACK 9a1e5e29317796c93bdef6c2df21ef50e6c68580
